File: advent_of_code/2024/python/13.py
import re
from typing import NamedTuple


class Point(NamedTuple):
    x: int
    y: int


def min_tokens(da, db, p):
    # da.x, db.x  *   na   =   p.x
    # da.y, db.y      nb       p.y

    D = da.x * db.y - da.y * db.x
    A = p.x  * db.y - p.y  * db.x
    B = da.x *  p.y - da.y *  p.x

    if D < 0:
        A *= -1
        B *= -1
        D *= -1

    if A % D == 0 and B % D == 0:
        return 3 * A // D + B // D
    
    return 0

def get_machines(text):
    machines = []
    for machine in text.split('\n\n'):
        button_a, button_b, prize = machine.splitlines()
        da = Point(*map(int, re.findall('\d+', button_a)))
        db = Point(*map(int, re.findall('\d+', button_b)))
        p = Point(*map(int, re.findall('\d+', prize)))
        machines.append((da, db, p))
    return machines

def part1(text):
    return sum(
        min_tokens(da, db, p)
        for da, db, p in get_machines(text)
    )

N = 10000000000000

def part2(text):
    return sum(
        min_tokens(da, db, Point(p.x+N,p.y+N))
        for da, db, p in get_machines(text)
    )

File: advent_of_code/2024/rust/src/13.rs
use regex::Regex;

#[derive(Debug, Clone, Copy)]
struct Point {
    x: i64,
    y: i64,
}

fn min_tokens(da: Point, db: Point, p: Point) -> i64 {
    let d = da.x * db.y - da.y * db.x;
    let mut a = p.x * db.y - p.y * db.x;
    let mut b = da.x * p.y - da.y * p.x;
    let mut d = d;

    if d < 0 {
        a = -a;
        b = -b;
        d = -d;
    }

    if a % d == 0 && b % d == 0 {
        return 3 * a / d + b / d;
    }

    0
}

fn get_machines(text: &str) -> Vec<(Point, Point, Point)> {
    let re = Regex::new(r"\d+").unwrap();

    text.split("\n\n")
        .filter_map(|machine| {
            let lines: Vec<&str> = machine.lines().collect();
            if lines.len() != 3 {
                return None;
            }

            let parse_point = |line: &str| -> Option<Point> {
                let mut numbers = re.find_iter(line).filter_map(|m| m.as_str().parse().ok());
                Some(Point {
                    x: numbers.next()?,
                    y: numbers.next()?,
                })
            };

            Some((
                parse_point(lines[0])?,
                parse_point(lines[1])?,
                parse_point(lines[2])?,
            ))
        })
        .collect()
}

fn part1(text: &str) -> i64 {
    get_machines(text)
        .into_iter()
        .map(|(da, db, p)| min_tokens(da, db, p))
        .sum()
}

const N: i64 = 10_000_000_000_000;

fn part2(text: &str) -> i64 {
    get_machines(text)
        .into_iter()
        .map(|(da, db, p)| {
            min_tokens(
                da,
                db,
                Point {
                    x: p.x + N,
                    y: p.y + N,
                },
            )
        })
        .sum()
}

fn main() {
    aoc2024::solve(part1, part2);
}
